### PR TITLE
Allow for arbitrary higher taxonomic levels in tax_unique()

### PR DESCRIPTION
**Pull request by willgearty**
This allows for users to specify arbitrary higher taxonomic levels (above genus, e.g. family, suborder, superclass) instead of being stuck with just family, order, and class. I've also moved some bits of code around to clean up the function. Finally, I also enhanced group_apply() to work with functions that have `...` arguments (like tax_unique() now does).

Fixes https://github.com/palaeoverse-community/palaeoverse/issues/52.

### CODE REVIEW

File: R/group_apply.R [hunk 1]
```diff
@@ -100,13 +100,15 @@ group_apply <- function(occdf, group, fun, ...) {
...
 
   supp_args <- list(...)
-  indx <- which(!(names(supp_args) %in% names(formals(fun))))
-  if (length(indx) > 1) {
-    stop(paste(paste0("`", names(supp_args)[indx], "`", collapse = "/"),
-               "are not valid arguments for the specified function"))
-  } else if (length(indx) == 1) {
-    stop(paste0("`", names(supp_args)[indx], "`",
-                " is not a valid argument for the specified function"))
+  if (!("..." %in% names(formals(fun)))) {
```
Comment (LewisAJones) [inline R/group_apply.R:103]: The updates with group_apply look good to me and probably something we should have implemented from the start! Nice job!

File: R/tax_unique.R [hunk 2]
```diff
@@ -16,12 +16,10 @@
...
 #' containing the genus-level identifications.
-#' @param family \code{character}. The name of the column in the dataframe
-#' containing the family-level identifications.
-#' @param order \code{character}. The name of the column in the dataframe
-#' containing the order-level identifications.
-#' @param class \code{character}. The name of the column in the dataframe
-#' containing the class-level identifications.
+#' @param ... \code{character}. Other named arguments specifying the names of
+#' columns of higher levels of taxonomy (e.g. subfamily, order, superclass).
+#' These arguments should be in ascending order from lowest to highest (see
```
Comment (LewisAJones) [inline R/tax_unique.R:21]:
```suggestion
#' These arguments should be in ascending order from lowest to highest taxonomic rank (see
```
Leave it or take it, up to you!

File: R/tax_unique.R [hunk 3]
```diff
@@ -135,105 +132,55 @@ tax_unique <- function(occdf = NULL, binomial = NULL, species = NULL,
...
-  if (is.null(family)) {
-    stop("`family` argument is missing. Provide column name of family names.")
-  }
-  if (!(family %in% colnames(occdf))) {
-    stop("`occdf` does not contain column name provided to `family`")
+  higher_args <- list(...)
+  higher_names <- names(higher_args)
+  higher_cols <- unname(unlist(higher_args))
+  if (length(higher_args) == 0) {
+    stop("At least one higher taxonomic level must be supplied (e.g. family)")
```
Comment (LewisAJones) [inline R/tax_unique.R:148]:
```suggestion
    stop("At least one higher taxonomic level must be supplied (e.g. `family`)")
```
Maybe put this in backticks similar to the other error messages to help indicate an additional argument is needed?

File: R/tax_unique.R [hunk 4]
```diff
@@ -135,105 +132,55 @@ tax_unique <- function(occdf = NULL, binomial = NULL, species = NULL,
...
-      "order"
-  }
-
-  if (!is.null(class) && !(class %in% colnames(occdf))) {
-    stop("`occdf` does not contain column name provided to `class`")
-  }
-  if (!is.null(class)) {
-    colnames(occdf)[colnames(occdf) == class] <-
-      "class"
+  for (level_label in names(higher_args)) {
```
Comment (LewisAJones) [inline R/tax_unique.R:150]:
```suggestion
  for (level_label in higher_names) {
```
Couldn't this just be replaced with higher_names from L145?

File: R/tax_unique.R [hunk 5]
```diff
@@ -135,105 +132,55 @@ tax_unique <- function(occdf = NULL, binomial = NULL, species = NULL,
...
+    col_name <- higher_args[[level_label]]
+    if (!(col_name %in% colnames(occdf))) {
+      stop(paste0("`occdf` does not contain column name provided to `",
+                  level_label, "`"))
+    }
+    #Substitute labels used in PBDB downloads
+    occdf[[col_name]] <-
+      gsub("NO_FAMILY_SPECIFIED|NO_ORDER_SPECIFIED|NO_CLASS_SPECIFIED", NA,
```
Comment (LewisAJones) [inline R/tax_unique.R:158]: Neat! Didn't realise you could use gsub with multiple values like this

**Review by LewisAJones — APPROVED** [review]
Hi Will,

I have very minor comments. All looks pretty good to me. The update with `group_apply` is useful and allows greater versatility beyond `tax_unique`, so good job! I think the updates with `tax_unique` have improved an already great function. It should allow the function to be used with a wider range of datasets. 

See above for my very minor comments.

If @bethany-j-allen agrees, I think you should also add yourself as a second developer.

Best,

Lewis

File: R/tax_unique.R (hunk 5 above)
Comment (willgearty) [reply R/tax_unique.R:158]: Yep, any regular expression works!

File: R/tax_unique.R [hunk 6]
```diff
@@ -16,12 +16,10 @@
...
 #' containing the genus-level identifications.
-#' @param family \code{character}. The name of the column in the dataframe
-#' containing the family-level identifications.
-#' @param order \code{character}. The name of the column in the dataframe
-#' containing the order-level identifications.
-#' @param class \code{character}. The name of the column in the dataframe
-#' containing the class-level identifications.
+#' @param ... \code{character}. Other named arguments specifying the names of
+#' columns of higher levels of taxonomy (e.g. subfamily, order, superclass).
+#' These arguments should be in ascending order from lowest to highest taxonomic rank (see
```
[the remark this answers is not in the document]
Comment (bethany-j-allen) [reply R/tax_unique.R:21]: I wonder whether we need to be even clearer than this, and literally say that the given order of the taxonomic variables is the order in which they are filtered.

**Review by bethany-j-allen — COMMENTED** [review]
Thanks a lot for doing this Will! My comments are mostly about making the inputs more understandable. Otherwise it looks great :)

**Comment by willgearty** [comment]
Thanks @bethany-j-allen, how does this look now?

**Comment by willgearty** [comment]
The most recent commit also fixes #50.

**Review by bethany-j-allen — APPROVED** [review]
Much more clearly explained I think! And thanks a lot for fixing the loophole. Looks ready to me.

**Comment by willgearty** [comment]
Thank you both, @bethany-j-allen and @LewisAJones!

**Pull request merged**